src/python/pants/backend/python/rules/python_create_binary.py

```diff
@@ -2,47 +2,77 @@
 # Licensed under the Apache License, Version 2.0 (see LICENSE).
 
 from dataclasses import dataclass
-from typing import Optional
+from typing import Optional, Tuple
 
 from pants.backend.python.rules.pex import Pex
 from pants.backend.python.rules.pex_from_target_closure import CreatePexFromTargetClosure
-from pants.backend.python.rules.targets import EntryPoint, PythonBinarySources
+from pants.backend.python.rules.targets import (
+    PexAlwaysWriteCache,
+    PexEmitWarnings,
+    PexIgnoreErrors,
+    PexIndices,
+    PexInheritPath,
+    PexRepositories,
+    PexShebang,
+    PexZipSafe,
+    PythonBinarySources,
+    PythonEntryPoint,
+    PythonPlatforms,
+)
 from pants.backend.python.targets.python_binary import PythonBinary
-from pants.build_graph.address import Address
 from pants.engine.addressable import Addresses
 from pants.engine.rules import UnionRule, rule
 from pants.engine.selectors import Get
-from pants.engine.target import Target
-from pants.rules.core.binary import BinaryImplementation, CreatedBinary
+from pants.rules.core.binary import BinaryConfiguration, CreatedBinary
 from pants.rules.core.determine_source_files import AllSourceFilesRequest, SourceFiles
 
 
 @dataclass(frozen=True)
-class PythonBinaryImplementation(BinaryImplementation):
-    required_fields = (EntryPoint, PythonBinarySources)
+class PythonBinaryConfiguration(BinaryConfiguration):
+    required_fields = (PythonEntryPoint, PythonBinarySources)
 
-    address: Address
     sources: PythonBinarySources
-    entry_point: EntryPoint
+    entry_point: PythonEntryPoint
 
-    # TODO: consume the other PythonBinary fields like `ZipSafe` and `AlwaysWriteCache`. These are
-    #  optional fields. If your target type has them registered, we can do extra meaningful things;
-    #  if you don't have them on your target type, we can still operate so long as you have the
-    #  required fields. Use `Target.get()` in the `create()` method.
+    always_write_cache: PexAlwaysWriteCache
+    emit_warnings: PexEmitWarnings
+    ignore_errors: PexIgnoreErrors
+    indices: PexIndices
+    inherit_path: PexInheritPath
+    repositories: PexRepositories
+    shebang: PexShebang
+    zip_safe: PexZipSafe
+    platforms: PythonPlatforms
 
-    @classmethod
-    def create(cls, tgt: Target) -> "PythonBinaryImplementation":
-        return cls(tgt.address, sources=tgt[PythonBinarySources], entry_point=tgt[EntryPoint])
+    def generate_additional_args(self) -> Tuple[str, ...]:
+        args = []
+        if self.always_write_cache.value is True:
+            args.append("--always-write-cache")
+        if self.emit_warnings.value is False:
+            args.append("--no-emit-warnings")
+        if self.ignore_errors.value is True:
+            args.append("--ignore-errors")
+        if self.indices.value is not None:
```

Comment:
We can easily fix that! What semantics do you think make sense?

```python
indices = []  # use `--no-index`
indices = ["foo"]  # pass via `--index`
indices = None  # don't tell Pex anything
```

NB that `indices = None` is equivalent to not specifying the field in the BUILD file.

Reply:
Yes - the semantics you describe is the only setup that seems to make sense. That said, think about supporting this at all once more. We do not need it for parity with v1 since v1 is broken here and the feature can't be being used. It may make sense to hold back these effectively new knobs at the target level until a need is shown. It would be a bit strange to have a single binary target that used a different set of indexes and find links from the rest of the repo.

Reply:
> We do not need it for parity with v1 since v1 is broken here

Correct that we don't necessarily need to consume the value. But, we do need to define and register the Field so that we don't make any changes to what is valid and invalid in BUILD files.

--

> It would be a bit strange to have a single binary target that used a different set of indexes and find links from the rest of the repo.

Perhaps, but this sounds feasible? Imagine a large org where the data science team wants to use some special index from Anaconda etc. but the majority of the org wants to stick with only PyPI.

Reply:
The unspoken detail is the devil.

That means that binary has 3rdparty dependencies declared locally in the same dir; since otherwise, if pulled from, say 3rdparty/python, that would mean there are repo-wide requirements there that don't work unless you switch indexes away from the repo-wide defaults.

If the local requirements are pulled from a requirements.txt file, we already support custom find links repos:
https://github.com/pantsbuild/pants/blob/b6a84251b78e5a917582a88ef6fff8287ccc702d/src/python/pants/backend/python/python_requirements.py#L47-L57

And if declared by hand in a BUILD file via `python_requirement_library`/`python_requirement`:
https://github.com/pantsbuild/pants/blob/b6a84251b78e5a917582a88ef6fff8287ccc702d/src/python/pants/python/python_requirement.py#L31

Both cases also probably ought to support custom indexes.

So I agree It does make sense for requirements to have configuration attached that says get me over here instead, but it probably does not make sense for a target depending on other targets and requirements to say this since the chances it knows what all its transtivie reqs are and where they might come from is doubtful.

Reply:
Thanks for explaining that all.

Here, would you propose then that we don't wire up this field at all to V2 binary? As mentioned, we do need to keep it for now as a registered field, but we signal it's unused in V2 and can deprecate the field.

Reply:
Going to merge to move forward on the project, but I'm happy to follow up on this. We can still change this behavior without the deprecation policy because it's V2.